### PR TITLE
navigator: support generic tail decoration for folder nodes

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/style/tree-decorators.css [hunk 1]
```diff
@@ -79,3 +79,4 @@
     bottom: 40%;
     right: 25%;
 }
+
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/tree-decorators.css]: We can revert this file, as the changes are unnecessary.

File: packages/core/src/browser/style/tree.css [hunk 2]
```diff
@@ -189,3 +189,4 @@
     cursor: default;
     opacity: var(--theia-mod-disabled-opacity);
 }
+
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/tree.css]: We can revert this file, as the changes are unnecessary.

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 3]
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
         }
-
         return icon;
     }
 
+    /**
+     * Render the tree node tail decorations.
+     * @returns the generic class icon used for the tail decoration
+     */
+    protected getGenericClassIconDecoration(): string[] {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: The documentation is misleading, the method does not do any rendering, it only returns the generic icon decoration.

```suggestion
    /**
     * Get the generic icon decoration.
     * @returns the class icon used for generic decorations.
     */
    protected getGenericClassIconDecoration(): string[] {
```

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 4]
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
+        return ['theia-marker-container-decoration', 'fa', 'fa-circle'];
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: I believe `theia-marker-container-decoration` is incorrect (this is provided by `@theia/markers` which `@theia/core` should not know about).

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 5]
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
-                    {content}
-                </div>;
-            })}
-        </React.Fragment>;
+        const decorationData = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+        const isCompositeTreeNode = CompositeTreeNode.is(node);
+
+        if (isCompositeTreeNode) {
+            // if the node is a composite, we just want to use the decorationData with the highest Priority (last one)
+            let tableauHighPrio: (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[] = [];
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: The naming of the variable is incorrect (we should stick to english as well).

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 6]
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
-                return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
-                    {content}
-                </div>;
-            })}
-        </React.Fragment>;
+        const decorationData = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+        const isCompositeTreeNode = CompositeTreeNode.is(node);
+
+        if (isCompositeTreeNode) {
+            // if the node is a composite, we just want to use the decorationData with the highest Priority (last one)
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]:
```suggestion
            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
```

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 7]
```diff
@@ -729,54 +729,113 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
+        //         const color = (decoration as TreeDecoration.TailDecorationIcon).color;
+        //         const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
+        //         const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
+        //         const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
+        //         const content = data ? data : icon ? <span key={node.id + 'icon' + index} className={this.getIconClass(icon)}></span> : '';
+        //         return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
+        //             {content}
+        //         </div>;
+        //     })}
+        // </React.Fragment>;
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: We should remove commented code.

File: packages/markers/src/browser/problem/problem-decorator.ts [hunk 8]
```diff
@@ -129,11 +129,19 @@ export class ProblemDecorator implements TreeDecorator {
...
             },
+            tailDecorations: [
+                {
+                    data: 'allo',
+                    color
+                }
+            ],
             iconOverlay: {
                 position,
                 icon,
```
Comment (vince-fugnitto) [inline packages/markers/src/browser/problem/problem-decorator.ts]: We should revert the changes when upstreaming.
Alternatively, we can have a separate commit which adds this data for test purposes (we can drop the commit when it would be ready to merge).

File: packages/scm/src/browser/decorations/scm-navigator-decorator.ts [hunk 9]
```diff
@@ -58,6 +58,7 @@ export class ScmNavigatorDecorator implements TreeDecorator {
                 }
             }
         }
+
```
Comment (vince-fugnitto) [inline packages/scm/src/browser/decorations/scm-navigator-decorator.ts]: We can revert this file, as the changes are unnecessary.

**Review by vince-fugnitto — COMMENTED** [review]
Please be sure to squash when ready and provide a meaningful commit title and message (it should also be in english).

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 10]
```diff
@@ -729,54 +729,82 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
-                </div>;
-            })}
-        </React.Fragment>;
+        const decorationData = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+
+        if (CompositeTreeNode.is(node)) {
+            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
+            // We put this decorationData in a new array so we can apply the map method on it
+            let arrayHighPrio: (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[] = [];
+            arrayHighPrio.push(decorationData.pop() as TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass);
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: The name of the variable does not mean much as is.
Also, why do we need to perform a map on a single element?

**Review by vince-fugnitto — COMMENTED** [review]
Please fix build errors:

```
@theia/core:   758:1   error  asterisks in jsdoc must be aligned (tslint:jsdoc-format)  @typescript-eslint/tslint/config
@theia/core:   776:17  error  'arrayHighPrio' is never reassigned. Use 'const' instead  prefer-const
@theia/core:   791:35  error  Missing semicolon   
```

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 11]
```diff
@@ -729,54 +729,82 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
         return icon;
     }
 
+    /**
+    * Get the generic icon decoration.
+     * @returns the class icon used for generic decorations.
+     */
+    protected getGenericClassIconDecoration(): string[] {
+        return ['fa', 'fa-circle'];
+    }
```
Comment (paul-marechal) [inline packages/core/src/browser/tree/tree-widget.tsx]: JSDoc looks weird.

File: packages/markers/src/browser/problem/problem-decorator.ts [hunk 12]
```diff
@@ -129,11 +129,19 @@ export class ProblemDecorator implements TreeDecorator {
         const icon = this.getOverlayIcon(marker);
         const color = this.getOverlayIconColor(marker);
         const priority = this.getPriority(marker);
+        //const isResource = Array.from(this.problemManager.getUris()).some(m => m === marker.uri);
+        //const problemStat = this.problemManager.getProblemStat();
```
Comment (vince-fugnitto) [inline packages/markers/src/browser/problem/problem-decorator.ts]: You can remove the following lines, it fails the build due to a missing whitespace.

```
lerna ERR! /home/evinfug/workspaces/omar-theia/packages/markers/src/browser/problem/problem-decorator.ts
lerna ERR!   132:9  error  Expected exception block, space or tab after '//' in comment  spaced-comment
lerna ERR!   133:9  error  Expected exception block, space or tab after '//' in comment  spaced-comment
```

Comment (vince-fugnitto) [reply packages/markers/src/browser/problem/problem-decorator.ts]: On a side note, you can enable [github actions](https://github.com/features/actions) on your repository through the settings to enable CI (so you can verify if your pull-requests build and tests pass):

image

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 13]
```diff
@@ -729,54 +729,75 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
-                return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
-                    {content}
-                </div>;
-            })}
-        </React.Fragment>;
+        const decorationData = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+
+        if (CompositeTreeNode.is(node)) {
+            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
+            const highPriorityDecoration = decorationData.pop();
```
Comment (paul-marechal) [inline packages/core/src/browser/tree/tree-widget.tsx]: This can be undefined, you should handle that case.

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 14]
```diff
@@ -763,14 +768,26 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
         return <React.Fragment>
-            {this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []).map((decoration, index) => {
+            {tailDecorations.map((decoration, index) => {
                 const { tooltip } = decoration;
                 const { data, fontData } = decoration as TreeDecoration.TailDecoration;
                 const color = (decoration as TreeDecoration.TailDecorationIcon).color;
-                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
                 const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
                 const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
+                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx:802]: minor: we can revert this change, it only moves the variable.

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 15]
```diff
@@ -763,14 +768,26 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
+        const tailDecorations = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+        if (tailDecorations.length === 0) {
+            return;
+        }
+        return this.renderTailDecorationsForNode(node, props, tailDecorations);
+    }
+
+    /**
+     * Render the tree node tail decorations if the node is not a `CompositeTreeNode`.
+     */
```
Comment (vince-fugnitto) [inline packages/core/src/browser/tree/tree-widget.tsx]: Documentation is misleading, we can remove it entirely.

File: packages/navigator/src/browser/navigator-widget.tsx [hunk 16]
```diff
@@ -139,6 +143,36 @@ export class FileNavigatorWidget extends FileTreeWidget {
...
+        if (tailDecorations.length === 0) {
+            return;
+        }
+
+        if (CompositeTreeNode.is(node)) {
+            return this.renderTailDecorationsForCompositeNode(node, props, tailDecorations);
+        } else {
+            return this.renderTailDecorationsForNode(node, props, tailDecorations);
+        }
+    }
```
Comment (vince-fugnitto) [inline packages/navigator/src/browser/navigator-widget.tsx]: We can make use of `FileStatNode` and rename the composite method:

```suggestion
        if (FileStatNode.is(node) && node.fileState.isDirectory) {
            return this.renderTailDecorationsForDirectoryNode(node, props, tailDecorations);
        } else {
            return this.renderTailDecorationsForNode(node, props, tailDecorations);
        }
    }
```

File: packages/navigator/src/browser/navigator-widget.tsx [hunk 17]
```diff
@@ -139,6 +143,36 @@ export class FileNavigatorWidget extends FileTreeWidget {
...
+
+    protected renderTailDecorationsForCompositeNode(node: TreeNode, props: NodeProps, tailDecorations:
+        (TreeDecoration.TailDecoration | TreeDecoration.TailDecorationIcon | TreeDecoration.TailDecorationIconClass)[]): React.ReactNode {
+        // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
```
Comment (vince-fugnitto) [inline packages/navigator/src/browser/navigator-widget.tsx]:
```suggestion
        // If the node represents a directory, we just want to use the decorationData with the highest priority (last element).
```
